docs/design/bin_pack_fault_domains.md

```diff
@@ -0,0 +1,126 @@
+          topologyKey: fault_domain
+        weight: 1
+    # AntiAffinity to not schedule the Pod on the same fault domain where other Pods with a different distribution key are running
+    podAntiAffinity:
+      preferredDuringSchedulingIgnoredDuringExecution:
+      - podAffinityTerm:
+          labelSelector:
+            matchExpressions:
+              key: "foundationdb.org/distribution-key"
+              operator: NotIn
+              values:
+              - "all-0"
+          topologyKey: fault_domain
+        weight: 1
+```
+
+The additional label and the `affinity` term enable the operator to spread Pods across fault domains.
+If `Required` is set to false this will be on a best effort base and can be useful if a fault domain is full but the Pod should be scheduled anyway.
+The following code snippet shows a possible implementation of the required structs:
+
+```golang
+type BinPack string
+
+const (
+    // BinPackAll defines that processes independent of their class should be bin packed
+    BinPackAll BinPack = "All"
+    // BinPackStateful defines that stateful processes should be bin packed and additionally stateless processes
+    BinPackStateful BinPack = "Stateful"
+    // BinPackClass only processes of the same class will be bin packed
+    BinPackClass BinPack = "Class"
+)
+
+// DistributionConfig
+type DistributionConfig struct {
+    // Enabled defines if the binpacking is enabled or not.
+    // Default: false
+    Enabled *bool
+    // DesiredFaultDomains desfines the number of desired fault domain.
+    // Should be greater than 0 is fault domain distribution is enabled.
+    // Default: 0
+    DesiredFaultDomains *int
+    // BinPack defines what processes should be bin packed to the fault domains.
+    // Default: "Class"
+    BinPack *BinPack
+    // Required defines if the affinity terms are required or preffered.
+    // Default: false
+    Required *bool
+}
+```
+
+The operator won't guarantee that the fault domains are equally used.
+Depending on the replacements for unreachable Pods we could use one fault domain more than another (or depending on the capacity),
+We also don't guarantee that we use exactly the number of `DesiredFaultDomains` e.g. when we spawn less processes than `DesiredFaultDomains` or when a fault domain is out of capacity and we are not using the `Required` with `true`.
+We also don't guarantee that the operator replaces any Pods that violate any constraint and we are not using the distribution with `Required` with `true` (`Required` with `true` would only ensure that we are not hurting the constraint).
+
+A change to `DesiredFaultDomains` will lead to a migration of all existing Pods in order to honor the new affinity term.
+
+### Reducing the number of DesiredFaultDomains
+
+If we currently use `DesiredFaultDomains` and we have `4` fault domains `f0`, `f1`, `f2` and `f3` with a Pod in each fault domain and we reduce `DesiredFaultDomains` to `2` this will result in the following steps:
```

Comment:
I worry that the term "fault domain" is overloaded here. Is it referring to the physical fault domains (i.e. the things that are used as the topology key in the affinity rules) or the logical fault domains (i.e. the things used as the match expression in the affinity rules)?

Reply:
In this case a logical fault domain == physical fault domain with the difference that a logical fault domain will be "bound" to a physical fault domain once the scheduler schedules the first Pod into the domain. I can add some more context on the difference between those in the design doc and maybe I come up with a better name.